src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -131,6 +153,72 @@ protected override void AddEmbeddedResourcesFromAddedModules(ArrayBuilder<Cci.Ma
         public override string Name => _metadataName;
         public AssemblyIdentity Identity => _sourceAssembly.Identity;
         public Version AssemblyVersionPattern => _sourceAssembly.AssemblyVersionPattern;
+
+        internal override SynthesizedAttributeData SynthesizeEmbeddedAttribute()
+        {
+            if ((object)_lazyEmbeddedAttribute != null)
+            {
+                return new SynthesizedAttributeData(
+                    _lazyEmbeddedAttribute.Constructor,
+                    ImmutableArray<TypedConstant>.Empty,
+                    ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.SynthesizeEmbeddedAttribute();
+        }
+
+        protected override SynthesizedAttributeData TrySynthesizeIsReadOnlyAttribute()
+        {
+            if ((object)_lazyIsReadOnlyAttribute != null)
+            {
+                return new SynthesizedAttributeData(
+                    _lazyIsReadOnlyAttribute.Constructor,
+                    ImmutableArray<TypedConstant>.Empty,
+                    ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.TrySynthesizeIsReadOnlyAttribute();
+        }
+
+        private void CreateEmbeddedAttributesIfNeeded()
+        {
+            if (_lazyEmbeddedAttributesDiagnostics.IsDefaultOrEmpty)
+            {
+                var diagnostics = DiagnosticBag.GetInstance();
+
+                if (this.NeedsGeneratedIsReadOnlyAttribute)
+                {
+                    CreateEmbeddedAttributeIfNeeded(
+                        ref _lazyEmbeddedAttribute,
+                        diagnostics,
+                        AttributeDescription.CodeAnalysisEmbeddedAttribute);
+
+                    CreateEmbeddedAttributeIfNeeded(
+                        ref _lazyIsReadOnlyAttribute,
+                        diagnostics,
+                        AttributeDescription.IsReadOnlyAttribute);
+                }
+
+                _lazyEmbeddedAttributesDiagnostics = diagnostics.ToReadOnlyAndFree();
+            }
+        }
+
+        private void CreateEmbeddedAttributeIfNeeded(ref SynthesizedEmbeddedAttributeSymbol symbol, DiagnosticBag diagnostics, AttributeDescription description)
+        {
+            if ((object)symbol == null)
```

Comment:
This is possibly going to be rewritten in Vlad's next PR, as there are more attributes to be generated. His feature will possibly need to generate some attributes that are generated in this PR, so the check will be necessary.
I see the future of embedded attributes going in one of two directions:
1) keeping the current design, having each feature calling `CreateEmbeddedAttributeIfNeeded` on the attributes it needs (possibly overlapping), and thus the current design protects against that.
2) If the features that need embedded attributes grow in numbers, there might be a need for an independent type manager to hold/manage them.